sdk/core/Azure.Core/src/Serialization/ModelWriter.cs

```diff
@@ -0,0 +1,193 @@
+        public ModelWriter(IModelJsonSerializable<object> model, ModelSerializerOptions options)
+        {
+            _model = model;
+            _options = options;
+        }
+
+        private SequenceBuilder GetSequenceBuilder()
+        {
+            if (_sequenceBuilder is null)
+            {
+                lock (_writeLock)
+                {
+                    if (_isDisposed)
+                    {
+                        throw new ObjectDisposedException(nameof(ModelWriter));
+                    }
+
+                    if (_sequenceBuilder is null)
+                    {
+                        SequenceBuilder sequenceBuilder = new SequenceBuilder();
+                        using var jsonWriter = new Utf8JsonWriter(sequenceBuilder);
+                        _model.Serialize(jsonWriter, _options);
+                        jsonWriter.Flush();
+                        _sequenceBuilder = sequenceBuilder;
+                    }
+                }
+            }
+            return _sequenceBuilder;
+        }
+
+        internal void CopyTo(Stream stream, CancellationToken cancellation)
+        {
+            SequenceBuilder builder = GetSequenceBuilder();
+            IncrementRead();
+            try
+            {
+                builder.CopyTo(stream, cancellation);
+            }
+            finally
+            {
+                DecrementRead();
+            }
+        }
+
+        internal bool TryComputeLength(out long length)
+        {
+            SequenceBuilder builder = GetSequenceBuilder();
+            IncrementRead();
+            try
+            {
+                return builder.TryComputeLength(out length);
+            }
+            finally
+            {
+                DecrementRead();
+            }
+        }
+
+        internal async Task CopyToAsync(Stream stream, CancellationToken cancellation)
+        {
+            SequenceBuilder builder = GetSequenceBuilder();
+            IncrementRead();
+            try
+            {
+                await builder.CopyToAsync(stream, cancellation).ConfigureAwait(false);
+            }
+            finally
+            {
+                DecrementRead();
+            }
+        }
+
+        /// <summary>
+        /// Converts the <see cref="ModelWriter"/> to a <see cref="BinaryData"/>.
+        /// </summary>
+        /// <returns>A <see cref="BinaryData"/> representation of the serialized <see cref="IModelJsonSerializable{T}"/> in JSON format.</returns>
+        public BinaryData ToBinaryData()
+        {
+            SequenceBuilder builder = GetSequenceBuilder();
+            IncrementRead();
+            try
+            {
+                bool gotLength = builder.TryComputeLength(out long length);
+                Debug.Assert(gotLength);
+                using var stream = new MemoryStream((int)length);
```

Comment:
Are we certain length will never be a long or should this cast go in a checked block?

Reply:
addressed here https://github.com/Azure/azure-sdk-for-net/pull/38137, but I opened this issue https://github.com/Azure/azure-sdk-for-net/issues/38141 to track if we want to add a helper since we will need to do this in many places.